In [425]:
import pandas as pd
import numpy as np
from math import sqrt

In [565]:
import warnings
warnings.filterwarnings("ignore", category=FutureWarning)

# CENTRAL TENDENCY

In [426]:
list=[1,2,3,4]

In [427]:
np.mean(list)

2.5

In [428]:
list1=[11,12,13,14,15]

In [429]:
np.median(list1)

13.0

In [430]:
list2=[10,11,12,13,14,15,10]

In [431]:
DATA= {
    "Name": ["FABINA", "HANEESHA", "MIKHTHAD", "SHIFNA", "RIDHA"],
    "Income": [50000, 65000, 70000, 85000, 90000],
    "Industry": ["DA","DS","DA","DS","DA"]
}

In [432]:
df = pd.DataFrame(DATA)
df

Name  Income Industry
0    FABINA   50000       DA
1  HANEESHA   65000       DS
2  MIKHTHAD   70000       DA
3    SHIFNA   85000       DS
4     RIDHA   90000       DA

In [433]:
df.Income.mean()

72000.0

In [434]:
np.mean(df["Income"])

72000.0

In [435]:
df.Income.median()

70000.0

In [436]:
np.median(df["Income"])

70000.0

In [437]:
df["Industry"].mode()

0    DA
Name: Industry, dtype: object

# Data Cleaning

## Filling the Null valueas

In [438]:
DATA= {
    "Name": ["FABINA", "HANEESHA", "MIKHTHAD", "SHIFNA", "RIDHA"],
    "Income": [50000, 65000, np.NaN, 85000, 90000],
    "Industry": ["DA","DS","DA","DS","DA"]
}

In [439]:
df = pd.DataFrame(DATA)
df

Name   Income Industry
0    FABINA  50000.0       DA
1  HANEESHA  65000.0       DS
2  MIKHTHAD      NaN       DA
3    SHIFNA  85000.0       DS
4     RIDHA  90000.0       DA

In [440]:
df.fillna(df["Income"].mean())

Name   Income Industry
0    FABINA  50000.0       DA
1  HANEESHA  65000.0       DS
2  MIKHTHAD  72500.0       DA
3    SHIFNA  85000.0       DS
4     RIDHA  90000.0       DA

In [441]:
Employees= {
    "Name": ["FABINA", "HANEESHA", "MIKHTHAD", "SHIFNA", "RIDHA","RASHID"],
    "Income": [50000, 65000, np.NaN, 85000, 90000,200000],
    "Industry": ["DA","DS","DA","DS","DA","HEAD"]
}

In [442]:
df=pd.DataFrame(Employees)

In [443]:
df

Name    Income Industry
0    FABINA   50000.0       DA
1  HANEESHA   65000.0       DS
2  MIKHTHAD       NaN       DA
3    SHIFNA   85000.0       DS
4     RIDHA   90000.0       DA
5    RASHID  200000.0     HEAD

In [444]:
np.mean(df["Income"])

98000.0

### Filling null values with mean

In [445]:
df.fillna(df["Income"].mean())

Name    Income Industry
0    FABINA   50000.0       DA
1  HANEESHA   65000.0       DS
2  MIKHTHAD   98000.0       DA
3    SHIFNA   85000.0       DS
4     RIDHA   90000.0       DA
5    RASHID  200000.0     HEAD

### Filling null values with mean

In [446]:
df.fillna(df["Income"].median())

Name    Income Industry
0    FABINA   50000.0       DA
1  HANEESHA   65000.0       DS
2  MIKHTHAD   85000.0       DA
3    SHIFNA   85000.0       DS
4     RIDHA   90000.0       DA
5    RASHID  200000.0     HEAD

In [447]:
df=pd.read_csv("housing.csv")

In [448]:
df

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0        -122.23     37.88                41.0        880.0           129.0   
1        -122.22     37.86                21.0       7099.0          1106.0   
2        -122.24     37.85                52.0       1467.0           190.0   
3        -122.25     37.85                52.0       1274.0           235.0   
4        -122.25     37.85                52.0       1627.0           280.0   
...          ...       ...                 ...          ...             ...   
20635    -121.09     39.48                25.0       1665.0           374.0   
20636    -121.21     39.49                18.0        697.0           150.0   
20637    -121.22     39.43                17.0       2254.0           485.0   
20638    -121.32     39.43                18.0       1860.0           409.0   
20639    -121.24     39.37                16.0       2785.0           616.0   

       population  households  median_income  median_house_value  \
0           322.0       126.0         8.3252            452600.0   
1          2401.0      1138.0         8.3014            358500.0   
2           496.0       177.0         7.2574            352100.0   
3           558.0       219.0         5.6431            341300.0   
4           565.0       259.0         3.8462            342200.0   
...           ...         ...            ...                 ...   
20635       845.0       330.0         1.5603             78100.0   
20636       356.0       114.0         2.5568             77100.0   
20637      1007.0       433.0         1.7000             92300.0   
20638       741.0       349.0         1.8672             84700.0   
20639      1387.0       530.0         2.3886             89400.0   

      ocean_proximity  
0            NEAR BAY  
1            NEAR BAY  
2            NEAR BAY  
3            NEAR BAY  
4            NEAR BAY  
...               ...  
20635          INLAND  
20636          INLAND  
20637          INLAND  
20638          INLAND  
20639          INLAND  

[20640 rows x 10 columns]

In [449]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [450]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity'],
      dtype='object')

In [451]:
np.mean(df["housing_median_age"])

28.639486434108527

In [452]:
np.median(df["housing_median_age"])

29.0

In [453]:
df["housing_median_age"].mode()

0    52.0
Name: housing_median_age, dtype: float64

In [454]:
np.mean(df["total_rooms"])

2635.7630813953488

In [455]:
np.median(df["total_rooms"])

2127.0

In [456]:
df["total_rooms"].mode()

0    1527.0
Name: total_rooms, dtype: float64

In [457]:
np.mean(df['total_bedrooms'])


537.8705525375618

In [458]:
np.median(df['total_bedrooms'])


nan

In [459]:
df["total_bedrooms"].mode()

0    280.0
Name: total_bedrooms, dtype: float64

In [460]:
df.population.mean()


1425.4767441860465

In [461]:
df.population.median()

1166.0

In [462]:
df.population.mode()

0    891.0
Name: population, dtype: float64

In [463]:
df.households.mean()

499.5396802325581

In [464]:
df.households.median()

409.0

In [465]:
df.households.mode()

0    306.0
Name: households, dtype: float64

In [466]:
df.median_income.mean()

3.8706710029069766

In [467]:
df.median_income.median()


3.5347999999999997

In [468]:
df.median_income.mode()


0     3.1250
1    15.0001
Name: median_income, dtype: float64

In [469]:
df.median_house_value.mean()

206855.81690891474

In [470]:
df.median_house_value.median()

179700.0

In [471]:
df.median_house_value.mode()


0    500001.0
Name: median_house_value, dtype: float64

In [472]:
df.ocean_proximity.mode()

0    <1H OCEAN
Name: ocean_proximity, dtype: object

# QUARTILE

In [473]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity'],
      dtype='object')

In [474]:
Q1=df.total_rooms.quantile(.25)
Q1

1447.75

In [475]:
Q2=df.total_rooms.quantile(.50)
Q2

2127.0

In [476]:
Q3=df.total_rooms.quantile(.75)
Q3

3148.0

In [477]:
df.households.quantile(.25)

280.0

In [478]:
df.households.quantile(.50)

409.0

In [479]:
df.households.quantile(.75)

605.0

In [480]:
df.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20640.000000  20640.000000        20640.000000  20640.000000   
mean    -119.569704     35.631861           28.639486   2635.763081   
std        2.003532      2.135952           12.585558   2181.615252   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1447.750000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.710000           37.000000   3148.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20640.000000  20640.000000   20640.000000   
mean       537.870553   1425.476744    499.539680       3.870671   
std        421.385070   1132.462122    382.329753       1.899822   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563400   
50%        435.000000   1166.000000    409.000000       3.534800   
75%        647.000000   1725.000000    605.000000       4.743250   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20640.000000  
mean        206855.816909  
std         115395.615874  
min          14999.000000  
25%         119600.000000  
50%         179700.000000  
75%         264725.000000  
max         500001.000000

In [481]:
df.total_rooms.quantile(.55)

2287.0

## create a function for mean,median,Q1,Q2,Q3 of a numerical values in a dataset.

In [482]:
def calculate_statastics(df):
    columns=[]
    mean = []
    median = []
    Q1 = []
    Q2 = []
    Q3 = []
    for i in df.select_dtypes(np.number):
        columns.append(i)
        mean.append(df[i].mean())
        median.append(df[i].median())
        Q1.append(df[i].quantile(.25))
        Q2.append(df[i].quantile(.50))
        Q3.append(df[i].quantile(.75))
    return (pd.DataFrame({"columns":columns,"mean":mean,"median":median,"Q1":Q1,"Q2":Q2,"Q3":Q3}))
   
data=pd.read_csv("housing.csv")
calculate_statastics(data)

columns           mean       median           Q1           Q2  \
0           longitude    -119.569704    -118.4900    -121.8000    -118.4900   
1            latitude      35.631861      34.2600      33.9300      34.2600   
2  housing_median_age      28.639486      29.0000      18.0000      29.0000   
3         total_rooms    2635.763081    2127.0000    1447.7500    2127.0000   
4      total_bedrooms     537.870553     435.0000     296.0000     435.0000   
5          population    1425.476744    1166.0000     787.0000    1166.0000   
6          households     499.539680     409.0000     280.0000     409.0000   
7       median_income       3.870671       3.5348       2.5634       3.5348   
8  median_house_value  206855.816909  179700.0000  119600.0000  179700.0000   

             Q3  
0    -118.01000  
1      37.71000  
2      37.00000  
3    3148.00000  
4     647.00000  
5    1725.00000  
6     605.00000  
7       4.74325  
8  264725.00000

In [483]:
data=[1,2,3,4,5,6,7,8,9,10]
df1=pd.DataFrame(data)

In [484]:
df1.quantile(.25)

0    3.25
Name: 0.25, dtype: float64

In [485]:
df1.quantile(.50)

0    5.5
Name: 0.5, dtype: float64

In [486]:
df1.quantile(.75)

0    7.75
Name: 0.75, dtype: float64

# STANDARD DEVIATION

###  A teacher wants to analyze the performance of two different classes (Class A and Class B) in a mathematics test graded out of 100 to determine which class has more consistent scores. The teacher has the following data:

### Class A Scores (out of 100): 78, 82, 85, 88, 90

### Class B Scores (out of 100): 55, 70, 85, 95, 100

In [487]:
CLASSA=[78,82,85,88,90]

In [488]:
DA=pd.DataFrame(CLASSA)

In [489]:
DA

0
0  78
1  82
2  85
3  88
4  90

In [490]:
np.std(DA)

C:\Users\HP\anaconda3\Lib\site-packages\numpy\core\fromnumeric.py:3643: FutureWarning: The behavior of DataFrame.std with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)
  return std(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)


0    4.270831
dtype: float64

In [491]:
CLASSB=[55, 70, 85, 95, 100]

In [492]:
DB=pd.DataFrame(CLASSB)

In [493]:
DB

0
0   55
1   70
2   85
3   95
4  100

In [494]:
np.std(DB)

C:\Users\HP\anaconda3\Lib\site-packages\numpy\core\fromnumeric.py:3643: FutureWarning: The behavior of DataFrame.std with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)
  return std(axis=axis, dtype=dtype, out=out, ddof=ddof, **kwargs)


0    16.552945
dtype: float64

## * Class A has a smaller standard deviation (4.27), indicating that the scores are more tightly grouped around the mean. This shows greater consistency in performance.
## * Class B has a larger standard deviation (16.55), meaning the scores are more spread out, reflecting less consistent performance.

# Coefficient of Variation
* The Coefficient of Variation (CV) is a statistical measure used to assess the relative variability of a dataset or a distribution. 
* It expresses the standard deviation as a percentage of the mean
* Instead of just looking at the absolute spread of values (like std), it looks at this spread relative to the size of the mean
* It is a standardized, unitless measure that allows you to compare variability between disparate groups and characteristics. It is also known as the relative standard deviation (RSD).


* compare the variability in SAT scores to ACT scores? While these college entrance exams are similar in nature and purpose, they use different scales. Consequently, you can’t compare their standard deviations directly.
* However, the coefficient of variation standardizes the raw data, which means you can compare the relative variability of SAT and ACT scores.

####  1.Two companies, A and B, report the monthly salaries of 5 employees each:
#### Company A salaries: [4800, 5200, 5000, 5100, 4900]
####  Company B salaries: [8000, 12000, 10000, 11000, 9000]
#### Which company has greater relative variability in salaries?

In [495]:
COMPANY_A_SALARIES=[4800,5200,5000,5100,4900]

In [496]:
np.std(COMPANY_A_SALARIES)

141.4213562373095

In [497]:
np.mean(COMPANY_A_SALARIES)

5000.0

In [498]:
cv=np.std(COMPANY_A_SALARIES)/np.mean(COMPANY_A_SALARIES)*100

In [499]:
cv

2.8284271247461903

### Company A: The CV is much smaller (2.83%), indicating that the salaries are highly consistent relative to the mean.

In [500]:
COMPANY_B_SALARIES=[8000,12000,10000,11000,9000]

In [501]:
np.std(COMPANY_B_SALARIES)

1414.213562373095

In [502]:
np.mean(COMPANY_B_SALARIES)

10000.0

In [503]:
cv=np.std(COMPANY_B_SALARIES)/np.mean(COMPANY_B_SALARIES)*100

In [504]:
cv

14.142135623730951

## Company B: The CV is significantly larger (14.14%), showing a greater relative variation in salaries.

## Salaries in Company A are more consistent, while Company B shows higher variability in employee salaries compared to the mean.

####  2.Two factories produce packets of sugar with the following weights (in grams):
####  factory X=[998,1002,1000,999,1001]
#### factoryY=[995,1005,990,1010,1000]
#### Which factory has more consistent packet weights?s?¶s?¶

In [505]:
factoryX=[998,1002,1000,999,1001]

In [506]:
np.mean(factoryX)

1000.0

In [507]:
np.std(factoryX)

1.4142135623730951

In [508]:
cv=np.std(factoryX)/np.mean(factoryX)*100
cv

0.14142135623730953

In [509]:
factoryY=[995,1005,990,1010,1000]

In [510]:
np.mean(factoryY)

1000.0

In [511]:
np.std(factoryY)

7.0710678118654755

In [512]:
cv=np.std(factoryY)/np.mean(factoryY)*100
cv

0.7071067811865476

# *Coefficient of Variation*

# Employee Performance:
### Question: Compare the variability in monthly sales (in $1000s) between two sales teams using CV. Which team has more consistent performance?
### Data:
### Team A: [50, 55, 48, 60, 53, 51, 59, 52]
### Team B: [70, 80, 75, 90, 85, 95, 72, 78]


In [513]:
Team_A=[50,55,48,60,53,51,59,52]


In [514]:
CVA=np.std(Team_A)/np.mean(Team_A)*100
CVA

7.417994330087637

In [515]:
Team_B=[70, 80, 75, 90, 85, 95, 72, 78]

In [516]:
CVB=np.std(Team_B)/np.mean(Team_B)*100
CVB

10.189007628308685

### Interpretation:
#### The CV represents the relative variability of sales as a percentage of the mean.

* Team A has a lower CV, indicating more consistent performance in monthly sales.
* Team B has a higher CV, suggesting greater variability in their sales performance.
#### Thus, Team A is more consistent in their monthly sales compared to Team B.

# Market Analysis:

### Question: Two stocks are being monitored for investment. Compare the variability in their weekly returns (%) using CV to determine which stock is more stable.
### Data:
### Stock A: [2.5, 2.7, 2.9, 2.6, 2.8, 3.0, 2.4]
### Stock B: [1.2, 1.8, 1.5, 1.7, 1.6, 1.4, 1.3]


In [517]:
Stock_A=[2.5, 2.7, 2.9, 2.6, 2.8, 3.0, 2.4]

In [518]:
Stock_B=[1.2, 1.8, 1.5, 1.7, 1.6, 1.4, 1.3]

In [519]:
CVA=np.std(Stock_A)/np.mean(Stock_A)*100
CVA

7.4074074074074066

In [520]:
CVB=np.std(Stock_B)/np.mean(Stock_B)*100
CVB

13.333333333333334

### Interpretation:
### The CV indicates the relative variability in weekly returns as a percentage of the mean:

* Stock A has a lower CV, showing more stability in its weekly returns.
* Stock B has a higher CV, suggesting greater variability and less stability in its returns.

### Stock A is the more stable investment option based on its consistent weekly returns compared to Stock B.

# Consumer Spending:

###	Question: Compare the variability in monthly grocery expenses (in $) between two households using CV to determine which household has more stable spending habits.
###	Data:
### Household A: [400, 420, 390, 410, 430, 400, 395]
### Household B: [200, 250, 230, 210, 240, 220, 260]

In [521]:
Household_A=[400, 420, 390, 410, 430, 400, 395]

In [522]:
Household_B=[200, 250, 230, 210, 240, 220, 260]

In [523]:
CVA=np.std(Household_A)/np.mean(Household_A)*100
CVA

3.2690817641016294

In [524]:
CVB=np.std(Household_B)/np.mean(Household_B)*100
CVB

8.695652173913043

### Interpretation:
* Household A has a much lower CV, indicating more consistent and stable monthly grocery spending.
* Household B shows a higher CV, suggesting greater variability in their monthly expenses.

### Household A demonstrates more stable spending habits compared to Household B.

# IQR-(INTER QUARTILE RANGE)

#### The IQR gives us the spread of data around the median and quantifies how much dispersion we have in the middle 50% of our distribution.
#### It can also be useful when checking the data for outliers. 
#### In addition, the IQR can be used to calculate a unitless measure of dispersion

## *  Lower Bound=Q1-1.5*IQR
## *  Upper Bound=Q3+1.5*IQR

In [525]:
df=pd.read_csv("weight-height.csv")

In [526]:
df

Gender     Height      Weight
0       Male  73.847017  241.893563
1       Male  68.781904  162.310473
2       Male  74.110105  212.740856
3       Male  71.730978  220.042470
4       Male  69.881796  206.349801
...      ...        ...         ...
9995  Female  66.172652  136.777454
9996  Female  67.067155  170.867906
9997  Female  63.867992  128.475319
9998  Female  69.034243  163.852461
9999  Female  61.944246  113.649103

[10000 rows x 3 columns]

In [527]:
meanH=np.mean(df["Height"])
meanH

66.36755975482124

In [528]:
meanW=np.mean(df["Weight"])
meanW

161.44035683283076

In [529]:
medianH=np.median(df["Height"])
medianH

66.31807008178464

In [530]:
medianW=np.median(df["Weight"])
medianW

161.21292769948298

In [531]:
def calculate_statastics(df):
    columns=[]
    mean = []
    median = []
    Q1 = []
    Q2 = []
    Q3 = []
    std = []          
    variance = []      
    CV = []           
    IQR = [] 
    for i in df.select_dtypes(np.number):
        columns.append(i)
        mean.append(df[i].mean())
        median.append(df[i].median())
        Q1.append(df[i].quantile(.25))
        Q2.append(df[i].quantile(.50))
        Q3.append(df[i].quantile(.75))
        std.append(df[i].std())  
        variance.append(df[i].var())  
        CV.append(df[i].std() / df[i].mean())
        IQR.append(df[i].quantile(0.75) - df[i].quantile(0.25)) 
    return (pd.DataFrame({"columns":columns,"mean":mean,"median":median,"Q1":Q1,"Q2":Q2,"Q3":Q3,"std": std,"Variance": variance,"CV": CV,"IQR": IQR}))
   
data=pd.read_csv("weight-height.csv")
calculate_statastics(data)

columns        mean      median          Q1          Q2          Q3  \
0  Height   66.367560   66.318070   63.505620   66.318070   69.174262   
1  Weight  161.440357  161.212928  135.818051  161.212928  187.169525   

         std     Variance        CV        IQR  
0   3.847528    14.803473  0.057973   5.668641  
1  32.108439  1030.951855  0.198887  51.351474

In [532]:
rangeW=(df["Weight"]).max()-(df["Weight"]).min()
rangeW

205.289571792353

In [533]:
rangeH=(df["Height"]).max()-(df["Height"]).min()
rangeH

24.735609021292504

In [534]:
VarianceH=np.var(df["Height"])
VarianceH

14.801992292876786

In [535]:
VarianceW=np.var(df["Weight"])
VarianceW

1030.8487602498426

In [536]:
SdH=np.std(df["Height"])
SdH

3.847335739557543

In [537]:
SdW=np.std(df["Weight"])
SdW

32.106833544431666

In [538]:
CvH=SdH/meanH*100
CvH

5.797012506969649

In [539]:
CvW=SdW/meanW*100
CvW

19.887736978726977

In [540]:
WQ1=(df["Weight"]).quantile(.25)
WQ1

135.8180513055015

In [541]:
WQ2=(df["Weight"]).quantile(.50)
WQ2

161.21292769948298

In [542]:
WQ3=(df["Weight"]).quantile(.75)
WQ3

187.16952486868348

In [543]:
WIQR=WQ3-WQ1
WIQR

51.35147356318197

In [544]:
weight_upper_bound=WQ3+1.5*WIQR
weight_upper_bound

264.19673521345646

In [545]:
weight_lower_bound=WQ1-1.5*WIQR
weight_lower_bound

58.79084096072856

In [546]:
HQ1=(df["Height"]).quantile(.25)
HQ1

63.505620481218955

In [547]:
HQ3=(df["Height"]).quantile(.75)
HQ3

69.1742617268347

In [548]:
HIQR=HQ3-HQ1
HIQR

5.668641245615746

In [549]:
height_upper_bound=HQ3+1.5*HIQR
height_upper_bound

77.67722359525831

In [550]:
height_lower_bound=HQ1-1.5*HIQR
height_lower_bound

55.00265861279534

In [551]:
out=df[(df['Height']<height_lower_bound)|(df['Height']>height_upper_bound)|(df['Weight']<weight_lower_bound)|(df['Weight']>weight_upper_bound)]
out

Gender     Height      Weight
994     Male  78.095867  255.690835
1317    Male  78.462053  227.342565
2014    Male  78.998742  269.989699
3285    Male  78.528210  253.889004
3757    Male  78.621374  245.733783
6624  Female  54.616858   71.393749
7294  Female  54.873728   78.606670
9285  Female  54.263133   64.700127

In [552]:
height=df["Height"][(df['Height']>height_lower_bound)&(df['Height']<height_upper_bound) ]
height

0       73.847017
1       68.781904
2       74.110105
3       71.730978
4       69.881796
          ...    
9995    66.172652
9996    67.067155
9997    63.867992
9998    69.034243
9999    61.944246
Name: Height, Length: 9992, dtype: float64

In [553]:
weight=df["Weight"][(df["Weight"]>weight_lower_bound)&(df["Weight"]<weight_upper_bound)]
weight

0       241.893563
1       162.310473
2       212.740856
3       220.042470
4       206.349801
           ...    
9995    136.777454
9996    170.867906
9997    128.475319
9998    163.852461
9999    113.649103
Name: Weight, Length: 9999, dtype: float64

In [554]:
df["height"]=height
df["weight"]=weight

In [555]:
df

Gender     Height      Weight     height      weight
0       Male  73.847017  241.893563  73.847017  241.893563
1       Male  68.781904  162.310473  68.781904  162.310473
2       Male  74.110105  212.740856  74.110105  212.740856
3       Male  71.730978  220.042470  71.730978  220.042470
4       Male  69.881796  206.349801  69.881796  206.349801
...      ...        ...         ...        ...         ...
9995  Female  66.172652  136.777454  66.172652  136.777454
9996  Female  67.067155  170.867906  67.067155  170.867906
9997  Female  63.867992  128.475319  63.867992  128.475319
9998  Female  69.034243  163.852461  69.034243  163.852461
9999  Female  61.944246  113.649103  61.944246  113.649103

[10000 rows x 5 columns]

In [556]:
df.isnull().sum()

Gender    0
Height    0
Weight    0
height    8
weight    1
dtype: int64

In [557]:
df

Gender     Height      Weight     height      weight
0       Male  73.847017  241.893563  73.847017  241.893563
1       Male  68.781904  162.310473  68.781904  162.310473
2       Male  74.110105  212.740856  74.110105  212.740856
3       Male  71.730978  220.042470  71.730978  220.042470
4       Male  69.881796  206.349801  69.881796  206.349801
...      ...        ...         ...        ...         ...
9995  Female  66.172652  136.777454  66.172652  136.777454
9996  Female  67.067155  170.867906  67.067155  170.867906
9997  Female  63.867992  128.475319  63.867992  128.475319
9998  Female  69.034243  163.852461  69.034243  163.852461
9999  Female  61.944246  113.649103  61.944246  113.649103

[10000 rows x 5 columns]

In [558]:
df.drop(["Height","Weight"],axis=1,inplace=True)


In [560]:
df

Gender     height      weight
0       Male  73.847017  241.893563
1       Male  68.781904  162.310473
2       Male  74.110105  212.740856
3       Male  71.730978  220.042470
4       Male  69.881796  206.349801
...      ...        ...         ...
9995  Female  66.172652  136.777454
9996  Female  67.067155  170.867906
9997  Female  63.867992  128.475319
9998  Female  69.034243  163.852461
9999  Female  61.944246  113.649103

[10000 rows x 3 columns]

In [566]:
df["height"].fillna(df['height'].mean(),inplace=True)

In [567]:
df["weight"].fillna(df['weight'].mean(),inplace=True)

In [568]:
df

Gender     height      weight
0       Male  73.847017  241.893563
1       Male  68.781904  162.310473
2       Male  74.110105  212.740856
3       Male  71.730978  220.042470
4       Male  69.881796  206.349801
...      ...        ...         ...
9995  Female  66.172652  136.777454
9996  Female  67.067155  170.867906
9997  Female  63.867992  128.475319
9998  Female  69.034243  163.852461
9999  Female  61.944246  113.649103

[10000 rows x 3 columns]

In [564]:
df.isnull().sum()

Gender    0
height    0
weight    0
dtype: int64

Scenario: 
An e-commerce company recently launched an email marketing campaign to promote a new product. The marketing team sent out 1000 emails and you are tasked with analysing how many recipients clicked on the link in the email(Success) vs how many did not(failure).
From the previous campaigns, you know that the avg Click Through Rate (CTR) is 15 %, meaning that in past campaigns 15% of the recipients typically click the link.
Calculate the probability of having exactly 150 recipients click to link in this campaign, given the campaign params as follows:

number of trials(email sent) n=1000
Prob of success=0.15
No. of success is k=150

In [2]:
import scipy.stats as stats
n=1000
p=0.15
k=150
p=stats.binom.pmf(k,n,p)

In [3]:
p

0.035310780109792626